File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        minimum, maximum = 0, 0

        for letter in s:
            if letter == '(':
                minimum += 1
                maximum += 1
            elif letter == ')':
                minimum -= 1
                maximum -= 1
            else:
                maximum += 1
                minimum -= 1
            if maximum < 0: return False
            minimum = max(minimum, 0)

        return minimum == 0

File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.java
class Solution {
    public boolean checkValidString(String s) {
        int min = 0, max = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                min++;
                max++;
            } else if (s.charAt(i) == ')') {
                min--;
                max--;
            } else {
                min--;
                max++;
            }
            if (max < 0) {
                return false;
            }
            min = Math.max(min, 0);
        }

        return min == 0;
    }
}